### PR TITLE
Optional fields

### PR DESCRIPTION
**Pull request by bobby1190**
Adding optional fields

### CODE REVIEW

File: spec/loaders/abstract-loader-shared-behavior.coffee [hunk 1]
```diff
@@ -303,6 +303,26 @@ registerSharedBehavior "AbstractLoaderSharedBehavior", (sharedContext) ->
...
+          loader.storageManager.storage('tasks').add([taskOne, taskTwo])
+
+        it 'returns false and does not call #_onLoadSuccess', ->
+          loader.setup(opts)
+          loader._checkCacheForData()
+          expect(loader._onLoadSuccess).not.toHaveBeenCalled()
+
+      context 'when optional fields have been requested but the fields are there on the tasks', ->
+        beforeEach ->
+          loader.storageManager.storage('tasks').add([taskOne, taskTwo])
```
Comment (cantino) [inline spec/loaders/abstract-loader-shared-behavior.coffee]: Don't we need to be doing something like `taskOne.set('test_field', 'fake value')` for both tasks in this context?

File: vendor/assets/javascripts/brainstem/loaders/abstract-loader.coffee [hunk 2]
```diff
@@ -185,7 +186,8 @@ class Brainstem.AbstractLoader
   _checkCacheForData: ->
     if @loadOptions.only?
       alreadyLoadedIds = _.select @loadOptions.only, (id) =>
-        @cachedCollection.get(id)?.associationsAreLoaded(@loadOptions.thisLayerInclude)
+        @cachedCollection.get(id)?.dependenciesAreLoaded(@loadOptions)
+        # and ensure any optional fields we need this time are in the cached models
```
Comment (cantino) [inline vendor/assets/javascripts/brainstem/loaders/abstract-loader.coffee]: remove this comment

File: vendor/assets/javascripts/brainstem/loaders/abstract-loader.coffee [hunk 3]
```diff
@@ -197,7 +199,8 @@ class Brainstem.AbstractLoader
 
       if cacheObject && cacheObject.valid
         subset = _.map cacheObject.results, (result) => @storageManager.storage(result.key).get(result.id)
-        if (_.all(subset, (model) => model.associationsAreLoaded(@loadOptions.thisLayerInclude)))
+        if (_.all(subset, (model) => model.dependenciesAreLoaded(@loadOptions)))
+          # if this if is still needed, we'd need to also check to make sure all optional_fields we need are in these models
```
Comment (cantino) [inline vendor/assets/javascripts/brainstem/loaders/abstract-loader.coffee]: remove this comment

File: vendor/assets/javascripts/brainstem/loaders/abstract-loader.coffee [hunk 4]
```diff
@@ -282,8 +285,9 @@ class Brainstem.AbstractLoader
     syncOptions.data.only = options.only.join(",") if options.only && @_shouldUseOnly()
     syncOptions.data.order = options.order if options.order?
     syncOptions.data.search = options.search if options.search
+    syncOptions.data.optional_fields = @loadOptions.optionalFields.join(",") if @loadOptions.optionalFields.length
```
Comment (cantino) [inline vendor/assets/javascripts/brainstem/loaders/abstract-loader.coffee]: Might want `@loadOptions.optionalFields?.length`

**Comment by cantino** [comment]
@jrolfs

**Comment by jrolfs** [comment]
This looks good.

File: spec/loaders/abstract-loader-shared-behavior.coffee [hunk 5]
```diff
@@ -333,6 +356,23 @@ registerSharedBehavior "AbstractLoaderSharedBehavior", (sharedContext) ->
...
+              taskOne.set('test_field', 'test value')
+
+            it 'calls #_onLoadSuccess with the models from the cache', ->
+              opts.optionalFields = ['test_field']
+              loader.setup(opts)
+              loader._checkCacheForData()
+              expect(loader._onLoadSuccess).toHaveBeenCalledWith([taskOne])
+
+          context 'all of the cached models do not have their optional fields loaded', ->
+            it 'returns false and does not call #_onLoadSuccess', ->
```
Comment (cantino) [inline spec/loaders/abstract-loader-shared-behavior.coffee]: Where do we test the return value?  Should that be here and in the other `it` block?

File: spec/loaders/abstract-loader-shared-behavior.coffee [hunk 6]
```diff
@@ -303,6 +303,29 @@ registerSharedBehavior "AbstractLoaderSharedBehavior", (sharedContext) ->
...
+          expect(loader._onLoadSuccess).not.toHaveBeenCalled()
+
+      context 'when optional fields have been requested and the fields are already on the tasks', ->
+        beforeEach ->
+          opts.optionalFields = ['test_field']
+          taskOne.set('test_field', 'fake value for one')
+          taskTwo.set('test_field', 'fake value for two')
+          loader.storageManager.storage('tasks').add([taskOne, taskTwo])
+
+        it 'returns true and does not calls #_onLoadSuccess', ->
```
Comment (cantino) [inline spec/loaders/abstract-loader-shared-behavior.coffee]: Don't see the return code being tested here either.

File: spec/loaders/abstract-loader-shared-behavior.coffee (hunk 5 above)
Comment (jrolfs) [inline spec/loaders/abstract-loader-shared-behavior.coffee]: Ideally, yes. The setup should go in a `beforeEach` and assertions in separate `it` blocks.

File: spec/loaders/abstract-loader-shared-behavior.coffee (hunk 6 above)
Comment (jrolfs) [inline spec/loaders/abstract-loader-shared-behavior.coffee]: Totally didn't read this description ಠ_ಠ. Same as below for testing the return value.

**Pull request merged**